Azure blog article (#15767)
pulumi/docs: content/blog/why-azure-resource-manager-templates-suck-for-deployments/index.md

If the tooling can’t deploy what you need, you're forced to write scripts and wire them in manually. That’s not DevOps—it’s duct tape.

**"Configuration management is all manual and custom."**

No secrets management, no config validation, no reusable environments, just fragile, homegrown scripts that break in subtle ways.

**"Our deployments are long and difficult. It all has to run sequentially."**

Because there's no dependency graph, no abstraction, and limited orchestration, deployments are slow and hard to parallelize.

---

## Pulumi: The Obvious Upgrade for .NET and Azure

Pulumi solves these problems at their root. It lets you define your Azure infrastructure using C#, the same language you're already using to build your applications. With Pulumi, you get:

✅ Familiar programming languages  
✅ Type safety and compile-time validation  
✅ Full IDE support (IntelliSense, refactoring, debugging)  
✅ Built-in support for Databricks, Kubernetes, and more  
✅ Automated config and secrets management  
✅ Easy testing, reuse, and CI/CD integration

Here’s a simple example: creating a Storage Account:

### ⛔ ARM Template (JSON)

```json

> {
>   "type": "Microsoft.Storage/storageAccounts",
>   "apiVersion": "2021-09-01",
>   "name": "[parameters('storageAccountName')]",
>   "location": "[resourceGroup().location]",
>   "sku": { "name": "Standard_LRS" },
>   "kind": "StorageV2",
>   "properties": {}
> }

Author: ```suggestion
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "storageAccountName": {
      "type": "string",
      "metadata": { "description": "Globally unique, 3–24 lowercase letters and numbers." }
    }
  },
  "resources": [
    {
      "type": "Microsoft.Storage/storageAccounts",
      "apiVersion": "2021-09-01",
      "name": "[parameters('storageAccountName')]",
      "location": "[resourceGroup().location]",
      "sku": { "name": "Standard_LRS" },
      "kind": "StorageV2",
      "properties": {}
    }
  ],
  "outputs": {
    "storageAccountId": {
      "type": "string",
      "value": "[resourceId('Microsoft.Storage/storageAccounts', parameters('storageAccountName'))]"
    }
  }
}
```

---

> title: "Why Azure Teams Are Moving from ARM Templates to Code"

Author: ```suggestion
title: "Why Azure Teams Are Moving from ARM Templates to .NET"
```

allow_long_title: true
date: 2025-08-22T01:41:10Z
draft: false

> meta_desc: ARM slowing you down? Ditch the JSON pain and deploy Azure like a pro with Pulumi + C#. Faster, cleaner, and actually developer-friendly.

Author: ```suggestion
meta_desc: ARM slowing you down? Ditch the JSON pain and deploy Azure like a pro with Pulumi + C#. Faster, cleaner, and actually developer-friendly.
meta_image: azure-resources-to-iac-general-code.png
```

authors:
    - sara-huddleston
tags:
    - azure
    - arm-templates
    - azure-resource-manager
---

[Azure Resource Manager (ARM)](https://www.pulumi.com/docs/iac/adopting-pulumi/migrating-to-pulumi/from-arm/) templates are powerful, but painful. If you’ve ever tried to provision cloud infrastructure using ARM, you know the challenges:

- Templates that started simple… and now span thousands of lines  
- Manual configuration stitched together with bespoke deployment logic  
- Lack of support for key services like Databricks  
- Slow, error-prone deployments that require multiple manual steps  
- No reuse, no testing, and no relief

> ARM wasn’t built for the complexity of modern Azure workloads. If you're already familiar with general-purpose languages, there’s a better path: Pulumi.

Author: ```suggestion
ARM wasn’t built for the complexity of modern Azure workloads. If you're already familiar with general-purpose languages, there’s a better path: [Pulumi](https://www.pulumi.com/docs/iac/clouds/azure/).
```

<!--more-->

> ## Top Pain Points of ARM Teamplates

Author: ```suggestion
## Top Pain Points of ARM Templates
```

            Kind = Kind.StorageV2
        });
    }
}
```

It’s clean. It’s familiar. And it works seamlessly with your .NET tooling.
Pulumi is open source, and your Pulumi Cloud account is free for individuals and small teams, with advanced editions for large enterprises. [Try Pulumi Cloud for FREE->](https://app.pulumi.com/signup)

## Why It Just Works for .NET Teams

If your organization is built on Azure and .NET:

- Use C# across both infrastructure and application layers
- Share libraries and logic between services
- Test infrastructure the same way you test code
- Integrate with Azure DevOps and GitHub Actions
- Manage secrets, environments, and policies as code

Pulumi supports all the Azure services ARM does (and more), while giving you flexibility and control ARM never could.

## Summary

[ARM templates ](https://www.pulumi.com/docs/iac/adopting-pulumi/migrating-to-pulumi/from-arm/)weren’t designed to scale with the complexity of today’s cloud environments. They’re static, verbose, hard to test, and increasingly brittle.

Pulumi gives you the tools to manage Azure the way you manage software: modular, testable, scalable, and secure.

> If you’re already building with .NET, you’re 90% of the way there. Why suffer through another slow, error-prone ARM deployment?

Author: ```suggestion
If you’re already building with C# and .NET, you’re 90% of the way there. Why suffer through another slow, error-prone ARM deployment?
```

{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "storageAccountName": {
      "type": "string",
      "metadata": { "description": "Globally unique, 3–24 lowercase letters and numbers." }
    }
  },
  "resources": [
    {
      "type": "Microsoft.Storage/storageAccounts",
      "apiVersion": "2021-09-01",
      "name": "[parameters('storageAccountName')]",
      "location": "[resourceGroup().location]",
      "sku": { "name": "Standard_LRS" },
      "kind": "StorageV2",
      "properties": {}
    }
  ],
  "outputs": {
    "storageAccountId": {
      "type": "string",
      "value": "[resourceId('Microsoft.Storage/storageAccounts', parameters('storageAccountName'))]"
    }
  }
}
```

### ✅ Azure with Pulumi in C#

```csharp
using Pulumi;
using Pulumi.AzureNative.Storage;

class MyStack : Stack
{
    public MyStack()
    {
        var storageAccount = new StorageAccount("mystorage", new StorageAccountArgs
        {
            ResourceGroupName = "my-rg",
            Sku = new SkuArgs { Name = SkuName.Standard_LRS },
            Kind = Kind.StorageV2
        });
    }
}
```

> It’s clean. It’s familiar. And it works seamlessly with your .NET tooling.
> Pulumi is open source, and your Pulumi Cloud account is free for individuals and small teams, with advanced editions for large enterprises. [Try Pulumi Cloud for FREE->](https://app.pulumi.com/signup)

Author: ```suggestion
It’s clean. It’s familiar. And it works seamlessly with your .NET tooling.
Pulumi is open source, and your Pulumi Cloud account is free for individuals and small teams, with advanced editions for large enterprises. 

[Try Pulumi Cloud for FREE->](https://app.pulumi.com/signup)
```

---

> title: "Why ARM Sucks for Azure Deployments (and What to Do About It)"
> allow_long_title: true

Author: ```suggestion
title: "Why Azure Teams Are Moving from ARM Templates to Pulumi"
allow_long_title: true
```

> date: 2025-08-21T01:41:10Z

Author: ```suggestion
date: 2025-08-22T01:41:10Z
```

**"Our deployments are long and difficult. It all has to run sequentially."**

Because there's no dependency graph, no abstraction, and limited orchestration, deployments are slow and hard to parallelize.

---

## Pulumi: The Obvious Upgrade for .NET and Azure

Pulumi solves these problems at their root. It lets you define your Azure infrastructure using C#, the same language you're already using to build your applications. With Pulumi, you get:

✅ Familiar programming languages  
✅ Type safety and compile-time validation  
✅ Full IDE support (IntelliSense, refactoring, debugging)  
✅ Built-in support for Databricks, Kubernetes, and more  
✅ Automated config and secrets management  
✅ Easy testing, reuse, and CI/CD integration

Here’s a simple example: creating a Storage Account:

### ⛔ ARM Template (JSON)

```json
{
  "type": "Microsoft.Storage/storageAccounts",
  "apiVersion": "2021-09-01",
  "name": "[parameters('storageAccountName')]",
  "location": "[resourceGroup().location]",
  "sku": { "name": "Standard_LRS" },
  "kind": "StorageV2",
  "properties": {}
}

> ```

Reviewer: @SaraDPH The Pulumi sample works as-is.
